src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs

```diff
@@ -268,24 +268,31 @@ private void UpdateEndpoints()
                             allParameterPolicies = MvcEndpointInfo.BuildParameterPolicies(routePattern.Parameters, _parameterPolicyFactory);
                         }
 
+                        // Replace parameter with literal value
                         var parameterRouteValue = action.RouteValues[parameterPart.Name];
 
-                        // Replace parameter with literal value
-                        if (allParameterPolicies.TryGetValue(parameterPart.Name, out var parameterPolicies))
+                        // Route value could be null if it is a "known" route value
+                        // Do not use the null value to de-normalize the route pattern
+                        // e.g. RouteValues will contain a null "page" value if there are Razor pages
+                        // Do not use the null "page" value to replace a {page} parameter
+                        if (parameterRouteValue != null)
                         {
-                            // Check if the parameter has a transformer policy
-                            // Use the first transformer policy
-                            for (var k = 0; k < parameterPolicies.Count; k++)
+                            if (allParameterPolicies.TryGetValue(parameterPart.Name, out var parameterPolicies))
                             {
-                                if (parameterPolicies[k] is IOutboundParameterTransformer parameterTransformer)
+                                // Check if the parameter has a transformer policy
+                                // Use the first transformer policy
+                                for (var k = 0; k < parameterPolicies.Count; k++)
                                 {
-                                    parameterRouteValue = parameterTransformer.TransformOutbound(parameterRouteValue);
-                                    break;
+                                    if (parameterPolicies[k] is IOutboundParameterTransformer parameterTransformer)
+                                    {
+                                        parameterRouteValue = parameterTransformer.TransformOutbound(parameterRouteValue);
+                                        break;
+                                    }
                                 }
                             }
-                        }
 
-                        segmentParts[j] = RoutePatternFactory.LiteralPart(parameterRouteValue);
+                            segmentParts[j] = RoutePatternFactory.LiteralPart(parameterRouteValue);
```

Comment:
Is there a unit test for this case? You could unit test this by creating an AD with `page = null` in required values

Reply:
Done